src/ui/utils.js

```diff
@@ -81,14 +82,28 @@ export function getBalloonTablePositionData( editor ) {
  * @returns {module:utils/dom/position~Options}
  */
 export function getBalloonCellPositionData( editor ) {
-	// This is a bit naive. See https://github.com/ckeditor/ckeditor5/issues/6357.
-	const modelTableCell = getTableCellAtPosition( editor.model.document.selection.getFirstPosition() );
-	const viewTableCell = editor.editing.mapper.toViewElement( modelTableCell );
+	const mapper = editor.editing.mapper;
+	const domConverter = editor.editing.view.domConverter;
+	const ranges = Array.from( editor.model.document.selection.getRanges() );
+
+	if ( ranges.length > 1 ) {
+		return {
+			target: () => createBoundingRect( ranges, modelRange => {
+				const modelTableCell = getTableCellAtPosition( modelRange.start );
+				const viewTableCell = mapper.toViewElement( modelTableCell );
+				return new Rect( domConverter.viewToDom( viewTableCell ) );
+			} ),
+			positions: BALLOON_POSITIONS
+		};
+	} else {
+		const modelTableCell = getTableCellAtPosition( ranges[ 0 ].start );
+		const viewTableCell = mapper.toViewElement( modelTableCell );
 
-	return {
-		target: editor.editing.view.domConverter.viewToDom( viewTableCell ),
-		positions: BALLOON_POSITIONS
-	};
+		return {
+			target: domConverter.viewToDom( viewTableCell ),
```

Comment:
Might use `getRectForTableCell` (`mapFn`) here AFAICS (to check) - two lines less :D

Reply:
Single range selection returns DOM element instead of function returning `Rect`.